src/distance.rs

```diff
@@ -72,9 +72,58 @@ pub fn pairwise<A: Float + Zero + AddAssign>(
     distances
 }
 
+#[derive(Default, Clone, Debug, Eq, PartialEq)]
+pub struct Cosine {}
+
+unsafe impl Sync for Cosine {}
+
+impl<A> Metric<A> for Cosine
+where
+    A: Float + AddAssign + std::iter::Sum,
+{
+    /// Cosine distance metric.
+    fn distance(&self, x1: &ArrayView1<A>, x2: &ArrayView1<A>) -> A {
+        self.rdistance(x1, x2).sqrt()
```

Comment:
`rdistance` computes the cosine distance directly, and thus taking the square root of the cosine distance is not meaningful.

In Euclidean distance:
- `rdistance` computes the squared distance (avoids `sqrt` for comparisons).
- `distance` computes the true distance by taking the square root of `rdistance`.

For cosine distance, there is no such squared relationship. The `rdistance` here is already the true cosine distance. The `.sqrt()` is mistakenly applying Euclidean logic to a non-Euclidean metric.

Please remove the `.sqrt()` call to align with the mathematical definition of cosine distance.

Reply:
According to https://en.wikipedia.org/wiki/Cosine_similarity, `cosine distance = (1 - cosine similarity).sqrt()` but `cosine distance = 1 - cosine similarity` is used generally. And since metric has `distance` and `rdistance`, I coded as such. That being said, I have no problem modifying it according to the general definition